src/Microsoft.AspNetCore.Mvc.TagHelpers/FormActionTagHelper.cs

```diff
@@ -0,0 +1,216 @@
+        /// </summary>
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        protected IUrlHelperFactory UrlHelperFactory { get; }
+
+        /// <summary>
+        /// The name of the action method.
+        /// </summary>
+        [HtmlAttributeName(ActionAttributeName)]
+        public string Action { get; set; }
+
+        /// <summary>
+        /// The name of the controller.
+        /// </summary>
+        [HtmlAttributeName(ControllerAttributeName)]
+        public string Controller { get; set; }
+
+        /// <summary>
+        /// The name of the area.
+        /// </summary>
+        [HtmlAttributeName(AreaAttributeName)]
+        public string Area { get; set; }
+
+        /// <summary>
+        /// Name of the route.
+        /// </summary>
+        /// <remarks>
+        /// Must be <c>null</c> if <see cref="Action"/> or <see cref="Controller"/> is non-<c>null</c>.
+        /// </remarks>
+        [HtmlAttributeName(RouteAttributeName)]
+        public string Route { get; set; }
+
+        /// <summary>
+        /// Additional parameters for the route.
+        /// </summary>
+        [HtmlAttributeName(RouteValuesDictionaryName, DictionaryAttributePrefix = RouteValuesPrefix)]
+        public IDictionary<string, string> RouteValues
+        {
+            get
+            {
+                if (_routeValues == null)
+                {
+                    _routeValues = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                return _routeValues;
+            }
+            set
+            {
+                _routeValues = value;
+            }
+        }
+
+        /// <inheritdoc />
+        /// <remarks>Does nothing if user provides an <c>formaction</c> attribute.</remarks>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if <c>formaction</c> attribute is provided and <see cref="Action"/>, <see cref="Controller"/>,
+        /// or <see cref="Route"/> are non-<c>null</c> or if the user provided <c>asp-route-*</c> attributes.
+        /// Also thrown if <see cref="Route"/> and one or both of <see cref="Action"/> and <see cref="Controller"/>
+        /// are non-<c>null</c>
+        /// </exception>
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            // If "formaction" is already set, it means the user is attempting to use a normal button or input element.
+            if (output.Attributes.ContainsName(FormAction))
+            {
+                if (Action != null || Controller != null || Area != null || Route != null || RouteValues.Count != 0)
```

Comment:
Good point. The `AnchorTagHelper` and `FormTagHelper` classes have the same issue.